device: rework `gen_handles.py`

JordanYates opened the pull request:
Rework the `gen_handles.py` script by pulling out all the elf parsing and dependency calculations into a dedicated file. This parent class will be used to progress #22545, as well as just making it simpler to add/change functionality in the future.

As an example of how this simplifies new features, the build system now attempts to generate a `.dot` file of the final device hierarchy. An example `dev_graph.dot` from `zephyr/tests/lib/devicetree/devices`:
![graph](https://user-images.githubusercontent.com/8476715/178131389-07d44425-34cf-4a07-b95a-9e6215e38cc3.png)

Commit:
pm: device: rename placeholder variable

Rename the placeholder variable generated for PM slots so that the
prefix doesn't colide with the PM structs declared by devices. This
simplifies the process of searching for symbols in `.elf` files.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
scripts: elf_parser: added

Adds a python module intended to simplify other scripts that need to
work with device information compiled into the first pass `.elf` file.

Scripts utilizing this module can focus on iterating over dependency
graphs to generate the desired output, instead of also needing to
extract and build the graphs in the first place.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
scripts: gen_handles: use `elf_parser.py`

Use the new elf_parser module to simplify the process of generating the
final device handle arrays.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
scripts: gen_handles: output dependency graph

Output the final dependency graph as a `.dot` file, which when rendered
by graphviz can be easier to comprehend than the text descriptions.

This output is optional in that it will not be generated if `graphviz`
is not installed.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
device: symetric injected dependencies

If a device manually specifies that it depends on a second DT device,
add the first device to the second devices list of supported devices.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
tests: devicetree: devices: test no label

Ensure that the build system works correctly for devices without a
nodelabel.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

gmarull commented on scripts/gen_handles.py:
```diff
@@ -127,6 +129,15 @@ def main():
+            dot = parsed_elf.device_dependency_graph('Device dependency graph', args.kernel)
+            with open(args.output_graphviz, 'w') as f:
+                f.write(dot.source)
+        except ImportError:
+            pass
```
not a fan of lazy imports, but users should probably get a message here, e.g. `Device dependency graph not available (missing graphviz)`.

JordanYates replied:
I don't think this warrants polluting build logs with an unconditional `print`.
Making it opt-in with `--verbose` or similar defeats the purpose, because if someone is enabling debug to see the error they would already know to install `graphviz`.

gmarull reviewed (dismissed):
LGTM, great work, one minor comment

Commit:
.gitignore: exempt `scripts/build` from ignore

Since the build scripts have been moved to `scripts/build`, exempt that
specific subdirectory from the `.gitignore` pattern.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

jeremybettis commented:
Did you really mean to have all these commits added to this PR? They don't seem to match the description of the PR.

JordanYates commented:
> Did you really mean to have all these commits added to this PR? They don't seem to match the description of the PR.

Yes, they are all related to the reworking of this script:
* Test no label: This case was causing build errors in initial versions of the graph generation
* rename placeholder: needed for the elf parsing to not pick up the placeholder as a device PM struct
* .gitignore: let git detect the changes to the scripts folder
* dependency graph: demonstrate new functionality from the rework
* symmetric dependencies: fixes an edge case with handle generation

The last one maybe is not 100% related to the PR, but it is 95% related.

mbolivar-nordic reviewed (commented):
A few changes from a quick skim.

I reject the assignee status as I don't own this script and I don't want to own this script :)

JordanYates commented on scripts/build/elf_parser.py:
```diff
@@ -0,0 +1,256 @@
+        self._ordinals_split.append(self._ordinals[prev:])
+
+    @property
+    def self_ordinal(self):
```
This one is mostly to be explicit about what ordinal it is accessing.
The ordinal of the device that created the array, not any of the dependency or support ordinals

JordanYates commented on scripts/build/elf_parser.py:
```diff
@@ -0,0 +1,256 @@
+        self.sym = sym
+        self.data = self.elf.symbol_data(sym)
+
+    def _self_native_read(self, offset):
```
Updated to `_data_native_read` to make it clearer that is is reading from the data member

mbolivar-nordic reviewed (commented):
ACK on the tests/lib/devicetree portions.

carlescufi commented:
@mbolivar-nordic could you +1 if happy?

galak reviewed (commented):
should .dot generation be optional and based on a Kconfig or VERBOSE setting?

JordanYates commented:
> should .dot generation be optional and based on a Kconfig or VERBOSE setting?

Its already "optional" in that if `graphviz` is not installed then the file isn't generated.
The file is broadly 2 lines of text per enabled device, which IMO is not significant enough in size to warrant disabling.
